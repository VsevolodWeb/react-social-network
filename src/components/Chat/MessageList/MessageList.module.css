.list {
    max-height: 170px;
    overflow: auto;
    margin-bottom: 20px;
}