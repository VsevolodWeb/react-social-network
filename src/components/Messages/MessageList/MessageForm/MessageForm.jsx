import React from 'react';
import {Field, reduxForm} from 'redux-form';

import { required, maxLength } from '../../../../utils/validators';
import { Textarea } from '../../../common/FormsControls/FormsControls';

const maxLength100 = maxLength(100);

const MessageForm = props => {
    return (
        <form onSubmit={props.handleSubmit}>
            <div className="form">
                <Field name="message" placeholder="Сообщение" component={Textarea} validate={[required, maxLength100]} />
                <div className="formGroup">
                    <button className="button">Отправить</button>
                </div>
            </div>
        </form>
    )
};

export default reduxForm({
    form: 'addMessage'
})(MessageForm);