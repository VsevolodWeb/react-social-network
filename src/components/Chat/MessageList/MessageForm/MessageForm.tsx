import React, {FC} from 'react'
import {InjectedFormProps, reduxForm} from 'redux-form'

import {maxLength, required} from '../../../../utils/validators'
import {CustomField, Textarea} from '../../../common/FormsControls/FormsControls'
import {MessageFormType} from '../MessageList'
import {Button} from 'antd'
import {WsChannelStatusType} from '../../Chat'

export type OwnPropsType = {
	wsChannelStatus: WsChannelStatusType | null
}

type PropsType = InjectedFormProps<MessageFormType, OwnPropsType> & OwnPropsType
type FieldNameType = 'message'

const maxLength100 = maxLength(100)

export const MessageForm: FC<PropsType> = props => (
	<form onSubmit={props.handleSubmit}>
		<div className="form">
			<CustomField<FieldNameType> name="message" placeholder="Сообщение" component={Textarea}
			                            validate={[required, maxLength100]}/>
			<div className="formGroup">
				<Button type="primary" htmlType="submit"
				        disabled={props.wsChannelStatus !== WebSocket.OPEN}>Отправить</Button>
			</div>
		</div>
	</form>
)

export default reduxForm<MessageFormType, OwnPropsType>({
	form: 'addMessage'
})(MessageForm)