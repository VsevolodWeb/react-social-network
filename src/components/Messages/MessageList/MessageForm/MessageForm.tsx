import React, {FC} from 'react';
import {Field, InjectedFormProps, reduxForm} from 'redux-form';

import { required, maxLength } from '../../../../utils/validators';
import {CustomField, Textarea} from '../../../common/FormsControls/FormsControls';
import {MessageFormType} from "../MessageList";

export type OwnPropsType = {}

type PropsType = InjectedFormProps<MessageFormType, OwnPropsType> & OwnPropsType

const maxLength100 = maxLength(100);

export const MessageForm: FC<PropsType> = props => (
    <form onSubmit={props.handleSubmit}>
        <div className="form">
            <Field name="message" placeholder="Сообщение" component={Textarea} validate={[required, maxLength100]} />
            <div className="formGroup">
                <button className="button">Отправить</button>
            </div>
        </div>
    </form>
);

export default reduxForm<MessageFormType, OwnPropsType>({
    form: 'addMessage'
})(MessageForm);