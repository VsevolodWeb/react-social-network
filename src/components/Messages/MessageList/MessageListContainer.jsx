import {connect} from 'react-redux';

import {addMessageActionCreator, messageChangeActionCreator} from '../../../redux/dialogs-reducer'
import MessageList from './MessageList';

const mapStateToProps = (state) => {
    return {
        dialog: state.dialog,
        newMessageValue: state.newMessageValue
    }
}

const mapDispatchToProps = (dispatch) => {
    return {
        addMessage: () => {
            dispatch(addMessageActionCreator(1));
        },
        updateNewPost: (value) => {
            dispatch(messageChangeActionCreator(value));
        }
    }
}

const MessageListContainer = connect(mapStateToProps, mapDispatchToProps)(MessageList)

export default MessageListContainer;