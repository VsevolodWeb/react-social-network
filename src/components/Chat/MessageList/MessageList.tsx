import React, {useEffect, useRef} from 'react'
import {useDispatch, useSelector} from 'react-redux'
import Message from './Message/Message'
import MessageForm from './MessageForm/MessageForm'
import {actions, sendMessage} from '../../../redux/chat-reducer'
import s from './MessageList.module.css'
import {getChatMessages} from '../../../redux/chat-selectors'

export type MessageFormType = {
	message: string
}

const MessageList = React.memo(props => {
	const listRef = useRef<HTMLDivElement>(null)
	const dispatch = useDispatch()
	const messages = useSelector(getChatMessages)

	let messagesElements = messages.map((item, index) => {
		return <Message data={item} key={index}/>
	})

	const addMessage = (data: MessageFormType) => {
		dispatch(sendMessage(data.message))
		dispatch(actions.resetMessage())
	}

	useEffect(() => {
		if (listRef.current) {
			listRef.current.scrollTop = listRef.current.scrollHeight
		}
	})

	return <>
		<div className={s.list} ref={listRef}>
			{messagesElements}
		</div>
		<MessageForm onSubmit={addMessage}/>
	</>
})

export default MessageList