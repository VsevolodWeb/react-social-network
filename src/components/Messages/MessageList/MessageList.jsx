import React from 'react';
import Message from './Message/Message';
import MessageForm from './MessageForm/MessageForm';


const MessageList = (props) => {
    let messageList = props.dialog.messages.map(item => {
        return <Message name={props.dialog.name} data={item} key={item.id} />
    });

    const addMessage = data => {
        props.addMessage(props.dialog.id, data.message);
    }

    messageList.push(
        <div key="0">
            <MessageForm onSubmit={addMessage} />
        </div>
    )

    return messageList
}

export default MessageList;