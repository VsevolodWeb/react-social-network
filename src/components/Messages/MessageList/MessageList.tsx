import React from 'react';
import Message from './Message/Message';
import MessageForm from './MessageForm/MessageForm';
import {DialogType} from "../../../redux/types/types";
import {actions} from "../../../redux/dialogs-reducer";

type PropsType = {
    dialog: DialogType
    addMessage: typeof actions.addMessage
    resetMessage: typeof actions.resetMessage
}

const MessageList: React.FC<PropsType> = props => {
    let messageList = props.dialog.messages.map(item => {
        return <Message name={props.dialog.name} data={item} key={item.id} />
    });

    const addMessage = data => {
        props.addMessage(props.dialog.id, data.message);
        props.resetMessage();
    };

    messageList.push(
        <div key="0">
            <MessageForm onSubmit={addMessage} />
        </div>
    );

    return messageList
};

export default MessageList;