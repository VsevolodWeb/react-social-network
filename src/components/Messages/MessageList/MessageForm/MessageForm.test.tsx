import React from 'react';
import renderer, {act} from 'react-test-renderer';
import { reduxForm } from 'redux-form';
import { createStore } from 'redux';
import { Provider } from 'react-redux';

import {MessageForm, OwnPropsType} from './MessageForm';
import {MessageFormType} from "../MessageList";

const store = createStore(() => ({}));
const Decorated = reduxForm<MessageFormType, OwnPropsType>({
    form: 'addMessage'
})(MessageForm);

describe('Products Component snapshot', () => {
    const component = renderer.create(
        <Provider store={store}>
            <Decorated/>
        </Provider>
    );

    it('show form', () => {
        const formElement = component!.root.findByType('form')
        let data;

        act(() => {
            data = formElement!.props.onSubmit();
        })

        console.log(data)

        expect(component).toMatchSnapshot()
    });
});