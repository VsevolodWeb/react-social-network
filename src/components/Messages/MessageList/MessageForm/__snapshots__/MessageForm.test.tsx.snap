// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Products Component snapshot show form 1`] = `
<form
  onSubmit={[Function]}
>
  <div
    className="form"
  >
    <div
      className="formGroup"
    >
      <textarea
        className="textElement"
        name="message"
        onBlur={[Function]}
        onChange={[Function]}
        onDragStart={[Function]}
        onDrop={[Function]}
        onFocus={[Function]}
        placeholder="Сообщение"
        rows={4}
        value=""
      />
    </div>
    <div
      className="formGroup"
    >
      <button
        className="button"
      >
        Отправить
      </button>
    </div>
  </div>
</form>
`;
